zkevm-circuits/src/blobkzg_circuit/src/main.rs

```diff
@@ -0,0 +1,585 @@
+/// x: evaluation point
+/// w: root of unity
+/// n: root of unity order
+#[derive(Clone, Debug)]
+struct InputValues<F: PrimeField> {
+    datapoints: [F; NUMBER_OF_POINTS],
+    x: F,
+    w: F,
+    n: F,
+}
+
+/// PRIVATE WITNESS VALUES FOR OUR CIRCUIT:
+/// evals: intermediate evaluation results
+/// wis: w^i for i in 0..4096 for evaluation
+/// xns: x^n intermediate exponentiation results
+/// x2is: x^(2^i) for i in 0...256 for exponentiation
+/// nbits: bits of n (4096) for exponentiation
+#[derive(Clone, Debug)]
+struct WitnessValues<F: PrimeField> {
+    evals: [F; 4098],
+    wis: [F; 4096],
+    xns: [F; 257],
+    x2is: [F; 256],
+    nbits: [F; 256],
+}
+
+impl<W: PrimeField, N: PrimeField> Default for BlobKZGCircuit<W, N> {
+    fn default() -> Self {
+        BlobKZGCircuit {
+            input_values: InputValues::<W>::default(),
+            witness_values: WitnessValues::<W>::default(),
+            _marker: std::marker::PhantomData,
+        }
+    }
+}
+
+impl<F: PrimeField> Default for InputValues<F> {
+    fn default() -> Self {
+        InputValues {
+            datapoints: [F::default(); NUMBER_OF_POINTS],
+            x: F::default(),
+            w: F::default(),
+            n: F::default(),
+        }
+    }
+}
+
+impl<F: PrimeField> Default for WitnessValues<F> {
+    fn default() -> Self {
+        WitnessValues {
+            evals: [F::default(); 4098],
+            wis: [F::default(); 4096],
+            xns: [F::default(); 257],
+            x2is: [F::default(); 256],
+            nbits: [F::default(); 256],
+        }
+    }
+}
+impl<W: PrimeField> WitnessValues<W> {
+    /// COMPUTATION OF PRIVATE WITNESS DATA
+    /// we have two main long computations:
+    /// a. the eval step
+    /// b. the xns step
+    ///
+    /// for the eval step we apply the main formula
+    /// to the datapoints, requiring the intermediate
+    /// rots of unity (wis) to be computed, as well as
+    /// the final exponentiation x^n (xns) to be computed.
+    ///
+    /// for the xns step we compute the square-and-multiply
+    /// algorithm, putting together all the possible powers
+    /// (x2is) to be computed and all the exponent's bits
+    /// (nbits) to be computed such that only the necessary powers
+    /// of x are multiplied together.
+    fn new(
+        InputValues {
+            datapoints,
+            x,
+            w,
+            n,
+        }: &InputValues<W>,
+    ) -> WitnessValues<W> {
+        let x2is: [W; 256] = {
+            let mut x2is = [W::ZERO; 256];
+            let mut state = *x;
+            for x in &mut x2is {
+                *x = state;
+                state *= state;
+            }
+            x2is
+        };
+        let nbits: [W; 256] = {
+            let mut nbits = [W::ZERO; 256];
+            let nbits_bool: [bool; 256] = {
+                let mut nbits_bool = [false; 256];
+                let repr: [u8; 32] = n.to_repr().as_ref().try_into().unwrap();
```

Comment:
We should consider failing here. the field binary representation will have an arbitrary bytes representation that won't necessarily fit 256. a safer option that won't be affected by endianness is to keep dividing by 2 and check of `n.is_odd` to set the bit on, until `n == 0` - this is, however, slower

Here, for instance, we can check 935. Given `as_ref` will yield big endian, we will iterate the two chunks [00000011, 10100111]. The final representation will be

[1100000011100101]

As we reverse the endianness here, we should probably add a comment stating so - that it is so to match the x2is powers and perform the square and add operation